modules/plugin/jdbc/jdbc-oracle/src/main/java/org/geotools/data/oracle/OracleDialect.java

```diff
@@ -617,23 +622,30 @@ OracleConnection unwrapConnection( Connection cx ) throws SQLException {
         }
         
         try {
-            // first lookup ever? (we have UNWRAPPER_NOT_FOUND as a sentinel for a lookup that
-            // will not work (we assume the datasource will always return connections we can
-            // unwrap, or never).
-            if (uw == null) {
-                UnWrapper unwrapper = DataSourceFinder.getUnWrapper(cx);
+            // Unwrap the connection multiple levels as necessary to get at the underlying
+            // OracleConnection. Maintain a map of UnWrappers to avoid searching
+            // the registry every time we need to unwrap.
+            Connection testCon = cx;
+            do {
+                UnWrapper unwrapper = uwMap.get(testCon.getClass());
                 if (unwrapper == null) {
-                    uw = UNWRAPPER_NOT_FOUND;
-                } else {
-                    uw = unwrapper;
+                    unwrapper = DataSourceFinder.getUnWrapper(testCon);
+                    if (unwrapper == null) {
+                        unwrapper = UNWRAPPER_NOT_FOUND;
+                    }
+                    uwMap.put(testCon.getClass(), unwrapper);
                 }
-            }
-            if (uw != null && uw != UNWRAPPER_NOT_FOUND) {
-                Connection uwcx = uw.unwrap( cx );
-                if ( uwcx != null && uwcx instanceof OracleConnection ) {
-                    return (OracleConnection) uwcx;
+                if (unwrapper == UNWRAPPER_NOT_FOUND) {
+                    // give up and try legacy approaches below
```

Comment:
The java 6 unwrapper is actually the "new" approach (the unwrappers were created when we were still java 4 compatible, and we never know which interfaces the env we're running into implements)